### ISSUE TITLE
Support of precompiled header behavior for squid sensor preprocessor

### ISSUE THREAD

**Issue by Bertk** [SonarOpenCommunity/sonar-cxx#118]
The analysis of larger projects can be improved if a precompiled header are simulated e.g. the specific header file is only read once and the result is preserved (cloned) for the further analysis.
Especially using boost shows a great slowdown of the analysis. I measured the following for a 330.000 LOC project:
```
INFO  -  * Sensors execution time breakdown:           206min 34s
INFO  -    o CxxSquidSensor:                           114min 10s (55%)
```
The portion for squid should be drastically reduced using simulating precompiled behavior.

**Comment by guwirth** [comment]
Did also a small test with a project with 18.000 LOC.

Without library header files:
```INI
sonar.cxx.includeDirectories=.
sonar.cxx.errorRecoveryEnabled=true
sonar.cxx.forceIncludes=VS10Macros.h
Total time: 22.410s, CxxSquidSensor done: 3784 ms
```
With library header files:
```INI
sonar.cxx.includeDirectories=.,C:/Program Files (x86)/Microsoft Visual Studio 10.0/VC/include,C:/Program Files (x86)/Microsoft Visual Studio 10.0/VC/atlmfc/include,C:/Program Files (x86)/Microsoft SDKs/Windows/v7.0A/Include,C:/Program Files (x86)/ETAS/EtasAdLib/Boost/1.49.0/include,C:/Program Files (x86)/ETAS/EtasAdLib/Xerces-C/3.1.1/include
sonar.cxx.errorRecoveryEnabled=true
sonar.cxx.forceIncludes=VS10Macros.h
Total time: 3:44.385s, Sensor CxxSquidSensor done: 208326 ms
```
* we have in total about 220 such projects, means **1h 20min** (220 x 22s) to **13h 42min** (220 x 3:44min).
* library header handling: 208326 ms - 3784 ms = 204s / project: total 220 x 204s = 12h 30min
* means maximum improvement with preprocessed header feature is 12h 20min

Conclusion: biggest win if we want to speed-up plugin

**Comment by wenns** [comment]
Thanks for this study. By the way: in my latest setups I usually skip including the system headers and external libs. If this causes parse problems, I inject macros via sonar.cxx.defines. I found this approach to be a viable alternative.

**Comment by guwirth** [comment]
[sonar-dev] Incremental read of source files
http://sonarqube.15.x6.nabble.com/sonar-dev-Incremental-read-of-source-files-td5032309.html

**Comment by Bertk** [comment]
The analysis performance should be improved to reduce cycle time for CI. The pre-compiled header support would be a big step forward because we use it in many projects.
I am not sure whether the SSLR tree can be serialized after "stdafx.h" is processed and restored if this was already done once.
@guwirth , @jmecosta Do you have a proposal for this feature?

**Comment by jmecosta** [comment]
Sharing my experience with precompile headers. Initially we started
enabling this per library, we did not see much improvement because we
always rebuild in our ci. Instead we created a single pch with system
headers + boost and some others that don't change much. All libraries were
setup with /use instead of /create. Builds times are half what they use to
be. The shortcoming is that the pch is huge, about 300mb. And parallelizing
the builds usually kills the machine if it does not have enough ram.

No idea how to implement this, however definitely something important
